### PR TITLE
Enable Translations

### PR DESCRIPTION
**Pull request by achow101**
I got most (nearly all I think) of the GUI strings that will need to be translated wrapped in the `tr()` function. I'm sure I missed a few, but I may have accidentally wrapped a few things that shouldn't be translated, so be sure to check for that during review.

I've also added the resource files for translations, but no code yet to handle actual translations. These resource files are from Transifex (except English) but I don't think the translations have been reviewed, so they may not be accurate. However, these translations won't be in use until languages can be selected.

Also, any new strings added that will appear in the GUI will need to be wrapped in the `tr()` function. `pylupdate4` will then need to be run to generate the English resource file and that file uploaded to Transifex.

### CODE REVIEW

**Comment by goatpig** [comment]
How does this work exactly? Do we have to implement a language selector in app?

**Comment by achow101** [comment]
At some point, yes we will need to implement a language selector. I'd put it in the settings. There will also need to be a change in the build process to add `lupdate` so that the language stuff will work. I'm looking into how doing all of this translation stuff should be done.

Also, getting the language selector isn't really urgent yet since no languages are close to being fully translated yet (after I tripled the number of strings to be translated).

**Comment by goatpig** [comment]
You could replace all entries with a lorem ipsum just for the sake of testing though. I'm more concerned about breaking some far fetched 6 layer deep dialog than the actual quality of completion of tsl.

**Comment by achow101** [comment]
well first we need to implement a language selector. RIght now, this breaks nothing. The only thing I am really concerned about are a few strings that had weird format stuff that I changed (not sure if I changed them correctly) and that I may have accidentally added something that wasn't a GUI string to be translated (e.g. an identifier for something in the wallet).

**Comment by goatpig** [comment]
At any rate the first is reviewing and testing these. Last time we one of the ATI devs tried to tr() everything in the Armory code base, it blew up in epic proportion and 99% of it was rolled back.

**Comment by achow101** [comment]
Right... There are actually a few strings that probably should have been left alone due to the weird formatting thing that was happening with them. I'll see if I can find them (qtdialogs is just massive) and revert them.

**Comment by goatpig** [comment]
Regardless this is a good time to do this change, right at the end of the early testing phase.

File: armorymodels.py [hunk 1]
```diff
@@ -1114,7 +1114,7 @@ def data(self, index, role=Qt.DisplayRole):
             return QVariant( cppAddr.getTxioCount())
          if col==COL.ChainIdx:
             if self.wlt.addrMap[addr160].chainIndex==-2:
-               return QVariant('Imported')
+               return QVariant(tr('Imported'))
```
Comment (achow101) [inline armorymodels.py]: Is this a label that is displayed or something for the wallet internally?

**Comment by achow101** [comment]
The string at https://github.com/achow101/BitcoinArmory/blob/355db40ad408008a407587de92838afe1529b493/qtdialogs.py#L4210 is questionable as to whether I formatted it correctly.

**Comment by droark** [comment]
utACK

No changes jumped out at me. Looks good. I feel like there was a reason Alan stopped using tr(), though. I can't remember why. I know there was an attempt at adding translations but it was pulled eventually. Anyone remember?

In any event, other than having to remember to wrap any new text in tr(), I'm fine with this commit. I'll test on OSX once it's finally running again.

**Comment by josephbisch** [comment]
utACK

Nothing jumped out at me either. However, I just scrolled through a lot of it, but not all.

The questionable string ([L4210 of qtdialogs.py](https://github.com/achow101/BitcoinArmory/blob/355db40ad408008a407587de92838afe1529b493/qtdialogs.py#L4210)) looks good to me, but I think the %s format specifiers will literally show up in the string to be translated, if I understand how this works. So the translators will need to know to leave the %s characters as is when they translate the string. The only other issue I see is that the use of %s to format the string leaves less flexibility for the translators. For example, what if theoretically there is a language that makes words plural in some way other than appending a character on the end of the word. But that is probably overthinking this and I don't know how you would handle that with PyQt.

**Comment by achow101** [comment]
I've replaced the `tr()`s with `self.tr()`s. I also squashed the commits.

**Comment by achow101** [comment]
It turns out `tr()` does work with the % formatting that Armory mostly uses for strings. I am now slowly going through the strings again and replacing them with QString's `.arg()` function.

**Comment by achow101** [comment]
Ok, I think I fixed everything and it's ready for review, and merging. I did some testing.

You can use https://github.com/achow101/BitcoinArmory/commit/c5b56e0217097b170279fac75ed633e2544b50a8 for testing. Just change the name of the .qm file to whatever other language you want to try translating. You will have to run `lrelease lang/*.ts` first.

**Comment by achow101** [comment]
I've finished everything that needs to be done for translations. You can now change the language in the settings.

All that's left is to wait for all of the translations to be completed. And testing and review of all of the changes.

**Comment by jimmysong** [comment]
@droark We were going to do the translations in a different module using python3's unicode support and Qt5. tr conflicted with too many things in Qt4, at least.

File: Makefile [hunk 2]
```diff
@@ -7,6 +7,7 @@ UNAME_S := $(shell uname -s)
 
 all:
 	$(MAKE) -C cppForSwig
+	lrelease lang/*.ts
```
Comment (achow101) [inline Makefile]: It is a tool that comes with Qt. It should be in the Qt stuff be default.

File: armorymodels.py [hunk 3]
```diff
@@ -111,7 +111,7 @@ def data(self, index, role=Qt.DisplayRole):
 
 
    def headerData(self, section, orientation, role=Qt.DisplayRole):
-      colLabels = ['', tr('ID'), tr('Wallet Name'), tr('Security'), tr('Balance')]
+      colLabels = ['', self.tr('ID'), self.tr('Wallet Name'), self.tr('Security'), self.tr('Balance')]
```
Comment (achow101) [inline armorymodels.py]: `self.tr` is a QObject function, so it will work if the class is a QObject.

**Comment by achow101** [comment]
I have addressed your comments (the unicode stuff in particular) and squashed and rebased.

**Comment by achow101** [comment]
Now I remember why those were put in unicode. Many languages don't have ASCII and some of the things that were put into unicode were because that part of the code would be reading things being displayed for some reason, but those were in other languages that didn't use ASCII and thus caused crashes.

File: ui/TxFrames.py [hunk 4]
```diff
@@ -1317,12 +1314,12 @@ def __init__(self, parent=None, main=None, initLabel=''):
       self.sentToSelfWarn = False
       self.fileLoaded = None
 
-      lblDescr = QRichLabel(\
+      lblDescr = QRichLabel(tr(
```
Comment (goatpig) [inline ui/TxFrames.py]: missing self.

File: ui/TxFrames.py [hunk 5]
```diff
@@ -1368,34 +1365,34 @@ def __init__(self, parent=None, main=None, initLabel=''):
...
-      self.infoLbls[-1].append(self.main.createToolTipWidget('The name of the wallet'))
-      self.infoLbls[-1].append(QRichLabel('<b>Wallet Label:</b>'))
+      self.infoLbls[-1].append(self.main.createToolTipWidget(self.tr('The name of the wallet')))
+      self.infoLbls[-1].append(QRichLabel(self.tr('<b>Wallet Label:</b>')))
       self.infoLbls[-1].append(QRichLabel(''))
 
       # ##
       self.infoLbls.append([])
-      self.infoLbls[-1].append(self.main.createToolTipWidget(\
+      self.infoLbls[-1].append(self.main.createToolTipWidget(tr(
```
Comment (goatpig) [inline ui/TxFrames.py]: missing self.

File: ui/TxFrames.py [hunk 6]
```diff
@@ -1729,29 +1725,29 @@ def signTx(self):
 
    def broadTx(self):
       if self.main.netMode == NETWORKMODE.Disconnected:
-         QMessageBox.warning(self, 'No Internet!', \
-            'Armory lost its connection to Bitcoin-Core, and cannot '
+         QMessageBox.warning(self, self.tr('No Internet!'), \
+            self.tr('Armory lost its connection to , and cannot '
```
Comment (goatpig) [inline ui/TxFrames.py]: Bitcoin-Core removed a whole in the few following sentences. No rephrasing though.

**Comment by goatpig** [comment]
Updated the code review

File: ui/MultiSigDialogs.py [hunk 7]
```diff
@@ -977,11 +975,11 @@ def createLockboxDashboardTab(self):
       # simulfunding widgets
       self.stkDashboard = QStackedWidget()
 
-      simultxt = 'Simul'
+      simultxt = self.tr('SimulFund')
```
Comment (achow101) [inline ui/MultiSigDialogs.py]: It came up in a transifex issue where the translator was saying that some of the strings say "check the Simul box" (or something like that) but Simul was not in the strings to be translated.

**Comment by achow101** [comment]
Updated with fixes.

File: ui/MultiSigDialogs.py (hunk 7 above)
Comment (goatpig) [inline ui/MultiSigDialogs.py]: It should be left as SimulFund then, without a tr encapsulation. Sadly I believe translating a trademark without the owner's approval is a form of trademark law violation.

Comment (achow101) [inline ui/MultiSigDialogs.py]: Ok. Done.

File: ArmoryQt.py [hunk 8]
```diff
@@ -3369,32 +3389,32 @@ def uriSendBitcoins(self, uriDict):
...
 
       selectedWalletID = None
       if len(self.walletMap)==0:
-         reply = QMessageBox.information(self, 'No Wallets!', \
-            'You just clicked on a "bitcoin:" link to send money, but you '
-            'currently have no wallets!  Would you like to create a wallet '
-            'now?', QMessageBox.Yes | QMessageBox.No)
+         reply = QMessageBox.information(self, self.tr('No Wallets!'), self.tr('''
+            You just clicked on a "bitcoin:" link to send money, but you 
+            urrently have no wallets!  Would you like to create a wallet
```
Comment (goatpig) [inline ArmoryQt.py]: typo

File: qtdialogs.py [hunk 9]
```diff
@@ -8389,12 +8367,10 @@ def wltTableClicked(self, currIndex, prevIndex=None):
 
       if not self.isBrowsingOnly:
          wlt = self.main.walletMap[self.selectedWltID]
-         self.btnSelectWlt.setText('%s Wallet: %s' % (self.actStr, self.selectedWltID))
+         self.btnSelectWlt.setText(self.tr('%1 Wallet: %2').arg(self.actStr, self.selectedWltID))
          nextAddr = wlt.peekNextUnusedAddr()
          
          addrStr = self.getAddrStr(wlt, nextAddr)
-         self.lblSelectWlt.setText('Will create new address: %s...' % \
```
Comment (achow101) [inline qtdialogs.py]: probably a bad rebase/squash. I have added it back in.

**Comment by achow101** [comment]
This has been merged as commits 03cef63a05f265fd3985c8ce5b373206a4b3fb18...f3353eb72c961108081aafacdb0c37bf8f9fcde9

**Pull request closed without merging**